### DISCUSSION TITLE
How to return 401 if user is not authenticated

### DISCUSSION THREAD

**Discussion by SlitiBrahim** [pocketbase/pocketbase#3002] [Q&A]
Hello, first, I want to thank you for the outstanding work you've been doing here!

I am currently encountering a problem that I'm not quite sure how to resolve:

I have a frontend Vue.js application (Nuxt, SSR) with protected routes, to which I only want to grant access to authenticated users. In order to keep users logged in after page reloads, I've established a store that contains their token. This store is persistent thanks to the use of `Pinia` and the `Pinia-plugin-persistedstate`.

Let's consider the following scenario:
I have a 'posts' collection that I only want authenticated users to be able to view, so I've set an API rule: @request.auth.id != "".

A user logs in, their token and user record are stored in the store and will remain accessible after page reloads. The user can fetch the posts and view them. However, after 2 years of inactivity, resulting in the user's token expiring, the user returns to my app. The browser retrieves the state (including the old token), and although the token has expired, the frontend app presumes the user is logged in and tries to fetch the posts. This results in a 200 response with an empty list, whereas what I want to happen is for the user to be logged out in this scenario. In other setups I'm familiar with, a 401 error is returned, causing the user to be logged out.

To resolve this issue, I understand there is the `await pb.collection('users').authRefresh()` method that checks the validity of the token, but should I execute this after every page load? The only method I've discovered to do this with Nuxt is to execute it in a middleware, but this implies it will be executed whenever I navigate from one page to another without refreshing the page within my SPA.

I have seen this issue here: [https://github.com/pocketbase/pocketbase/discussions/1611](https://github.com/pocketbase/pocketbase/discussions/1611)

Could you assist me in understanding how to set up a system like this with Pocketbase? Thank you so much.

**Comment by ganigeorgiev** [comment]
For the reasons why we return 200 on unsatisfied list rule, you can also check https://github.com/pocketbase/pocketbase/discussions/1570.

Note that even if you call `authRefresh()` it is not guaranteed that the authenticated user will satisfy the rules of the collection as they act also as filter.

Unfortunately I'm failing to understand where your application is running (client-side, server-side, both?) and I'm not sure what to recommend. You can also loosely check whether the user has expired token loaded in the AuthStore by calling `pb.authStore.isValid`.

**Reply by SlitiBrahim** [reply]
Hi! thanks for you quick answer.

The app is running on client-side, but I don't have access to the localStorage for example as the pages are rendered on the server-side (Nuxtjs SSR).

Does the `pb.authStore.isValid` only check if the token expiration date hasn't passed? This might not be entirely appropriate because it doesn't account for situations where the user no longer exists, such as if the user has deleted their account or an administrator has removed the user.

So my question is: On my frontend application, how can I check with a server-side check that the local token is still valid (user still exist, token didn't expire) and how should I get the latest user record (it can be that the user record has been updated like his home address for example and the frontend local record doesn't have this update)? Should I use the authRefresh() function or is there a better way? Basically how should I keep my users always logged in but have a mechanism that checks that they have a valid token otherwise they should be logged out?

I hope my questions are clear haha.. please do not hesitate to let me know if it's not the case.

Thank you so much

**Reply by ganigeorgiev** [reply]
>The app is running on client-side, but I don't have access to the localStorage for example as the pages are rendered on the server-side (Nuxtjs SSR).

So you have a mixed access to the PocketBase APIs - both from browser and the node server, right?
Usually when your app is only client-side SPA (aka. there is no SSR), then you can call `authRefresh()` on the first page load. When executed in a server-side context and you want to validate the authStore so that you can use it in your node actions to perform various checks, then you'll have to call `auhtRefresh()` on every request.


> how can I check with a server-side check that the local token is still valid (user still exist, token didn't expire) and how should I get the latest user record (it can be that the user record has been updated like his home address for example and the frontend local record doesn't have this update)? 

You can validate the current loaded authStore token and get the latest user record by calling `pb.collection('users').authRefresh()`.

An alternative approach if the View rule of your `users` collection is something like `@request.auth.id = id`, then you can send `pb.collection('users').getOne(pb.authStore.model?.id)` request. If it returns 404 than either the auth token was invalid or the user no longer exists.

> Basically how should I keep my users always logged in but have a mechanism that checks that they have a valid token otherwise they should be logged out?

PocketBase is fully stateless and doesn't have sessions in the traditional sense, so we really doesn't have "login" or "logout" actions. A user is considered "logged in" for a **single** request if they send `Authorization: TOKEN` header with it. This header is added automatically by the SDKs if there is a token available in `pb.authStore`.
In other words `pb.authStore` is a local persistent layer that holds the token and populate the `Authorization` header between requests. Each request operates separately and has their own separate context/state.

**Reply by SlitiBrahim** [reply]
Amazing! I see now. Thank you very much, I've just turned my frontend app to an SPA (and not SSR app), I grab the user token from the localStorage and then on every page navigation I call the `authRefresh` method. Thanks again!
